File: src/lei.ts
/**
 * LEI (Legal Entity Identifier).
 *
 * ISO 17442. 20-character alphanumeric code
 * identifying legal entities in financial
 * transactions. Validated using ISO 7064
 * Mod 97-10 (same as IBAN).
 *
 * @see https://www.gleif.org/
 */

import { mod97 } from "#checksums/mod97";
import { clean } from "#util/clean";
import { err } from "#util/result";
import { charValue, isalnum } from "#util/strings";

import type { ValidateResult, Validator } from "./types";
import { randomInt } from "#util/generate";

const compact = (value: string): string =>
  clean(value, " -").toUpperCase();

const validate = (value: string): ValidateResult => {
  const v = compact(value);
  if (v.length !== 20) {
    return err(
      "INVALID_LENGTH",
      "LEI must be exactly 20 characters",
    );
  }
  if (!isalnum(v)) {
    return err(
      "INVALID_FORMAT",
      "LEI must contain only letters and digits",
    );
  }
  // Characters 5-18 must be alphanumeric (entity)
  // Characters 19-20 are check digits (numeric)
  const checkPart = v.slice(18);
  if (!/^\d{2}$/.test(checkPart)) {
    return err(
      "INVALID_FORMAT",
      "LEI check digits must be numeric",
    );
  }
  // Convert to numeric and validate mod 97
  let numeric = "";
  for (let i = 0; i < v.length; i++) {
    const ch = v[i];
    if (ch !== undefined) {
      numeric += String(charValue(ch));
    }
  }
  if (mod97(numeric) !== 1) {
    return err(
      "INVALID_CHECKSUM",
      "LEI check digits are incorrect",
    );
  }
  return { valid: true, compact: v };
};

const format = (value: string): string => {
  const v = compact(value);
  return `${v.slice(0, 4)} ${v.slice(4, 8)} ${v.slice(8, 12)} ${v.slice(12, 16)} ${v.slice(16)}`;
};

/** Generate a random valid LEI. */
const generate = (): string => {
  for (;;) {
    let base = "";
    for (let i = 0; i < 18; i++) base += String(randomInt(0, 9));
    for (let cd = 0; cd < 100; cd++) {
      const c = base + String(cd).padStart(2, "0");
      if (validate(c).valid) return c;
    }
  }
};

/** Legal Entity Identifier. */
const lei: Validator = {
  name: "Legal Entity Identifier",
  localName: "Legal Entity Identifier",
  abbreviation: "LEI",
  entityType: "company",
  sourceUrl: "https://www.gleif.org/",
  examples: ["5493006MHB84DD0ZWV18"] as const,
  compact,
  format,
  validate,
  generate,
};

export default lei;
export { compact, format, validate, generate };

File: crates/stdnum-core/src/validators/global/lei.rs
use crate::{
  compact_without, random_digits,
  types::{
    EntityType, ValidationError, ValidationResult, Validator, ValidatorScope,
    ValidatorSpec,
  },
};

pub static VALIDATOR: Validator = Validator::new(ValidatorSpec {
  id: "lei",
  name: "Legal Entity Identifier",
  local_name: "Legal Entity Identifier",
  abbreviation: "LEI",
  aliases: &["LEI", "Legal Entity Identifier"],
  candidate_pattern: r"[A-Z0-9]{18}\d{2}",
  scope: ValidatorScope::Global,
  entity_type: EntityType::Company,
  source_url: Some("https://www.gleif.org/"),
  lengths: &[],
  examples: &["5493006MHB84DD0ZWV18"],
  compact,
  format,
  validate,
  generate: Some(generate),
  parse: None,
});

#[must_use]
pub fn compact(value: &str) -> String {
  compact_without(value, &[' ', '-']).to_uppercase()
}

fn mod97(value: &str) -> Option<u32> {
  let mut remainder = 0_u32;
  for ch in value.chars() {
    if ch.is_ascii_digit() {
      remainder = remainder
        .saturating_mul(10)
        .saturating_add(ch.to_digit(10)?)
        .rem_euclid(97);
    } else if ch.is_ascii_uppercase() {
      let numeric = u32::from(ch)
        .saturating_sub(u32::from('A'))
        .saturating_add(10);
      remainder = remainder
        .saturating_mul(100)
        .saturating_add(numeric)
        .rem_euclid(97);
    } else {
      return None;
    }
  }
  Some(remainder)
}

#[must_use]
pub fn format(value: &str) -> String {
  let value = compact(value);
  value
    .as_bytes()
    .chunks(4)
    .map(|chunk| std::str::from_utf8(chunk).unwrap_or(""))
    .collect::<Vec<_>>()
    .join(" ")
}

pub fn validate(value: &str) -> ValidationResult {
  let value = compact(value);
  if value.len() != 20 {
    return Err(ValidationError::InvalidLength(
      "LEI must be exactly 20 characters",
    ));
  }
  if !value.chars().all(|ch| ch.is_ascii_alphanumeric()) {
    return Err(ValidationError::InvalidFormat(
      "LEI must contain only letters and digits",
    ));
  }
  if !value
    .get(18..)
    .is_some_and(|part| part.chars().all(|ch| ch.is_ascii_digit()))
  {
    return Err(ValidationError::InvalidFormat(
      "LEI check digits must be numeric",
    ));
  }
  if mod97(&value) != Some(1) {
    return Err(ValidationError::InvalidChecksum(
      "LEI check digits are incorrect",
    ));
  }
  Ok(value)
}

#[must_use]
pub fn generate() -> String {
  loop {
    let base = random_digits(18);
    for check in 0..100 {
      let candidate = format!("{base}{check:02}");
      if validate(&candidate).is_ok() {
        return candidate;
      }
    }
  }
}
